test/unit/prof_small.c

```diff
@@ -0,0 +1,37 @@
+#include "test/jemalloc_test.h"
+
+TEST_BEGIN(test_profile_small_allocations) {
+	test_skip_if(!config_prof);
+
+	for (szind_t index = 0; index < SC_NBINS; index++) {
+		size_t size = sz_index2size(index);
+		assert_zu_le(size, SC_SMALL_MAXCLASS,
+		    "Unexpected large size class");
+		void *ptr = malloc(size);
+		assert_ptr_not_null(ptr, "Unexpected malloc failure");
+		edata_t *edata = emap_edata_lookup(TSDN_NULL,
+		    &arena_emap_global, ptr);
+		assert_ptr_not_null(edata,
+		    "Unable to find edata for allocation");
+		expect_false(edata_slab_get(edata),
+		    "Sampled small allocations should not be placed on slabs");
+		expect_ptr_eq(edata_base_get(edata), ptr,
+		    "Sampled allocations should be page-aligned");
+		expect_zu_eq(edata_usize_get(edata), size,
+		    "Edata usize did not match requested size");
+		expect_zu_eq(edata_size_get(edata), PAGE_CEILING(size),
+		    "Edata actual size was not a multiple of PAGE");
```

Comment:
It looks this is firing because of the extra page from cache_oblivious. We should fix it from this line:
https://github.com/jemalloc/jemalloc/blob/90176f8a87a0b5bdb0ac4c1a515b1d9c58dc5a82/src/arena.c#L337

Maybe this way:
```
size_t pad_size =  ? sz_large_pad : 0
size_t esize = usize + pad_size;
...
...
if (edata != NULL && pad_size != 0) {
    arena_cache_oblivious_randomize
```

Then change this branch into an assert:
https://github.com/jemalloc/jemalloc/blob/90176f8a87a0b5bdb0ac4c1a515b1d9c58dc5a82/include/jemalloc/internal/arena_inlines_b.h#L494

I believe this would fix the other assertion failure in demote as well

Reply:
I ran it locally with `--enable-cache-oblivious` and was able to reproduce the failure. Re-running after applying your suggested change does indeed fix it.

Reply:
Have to add the missing assert still, so hold off on reviewing again

Reply:
Now we're getting failures on the realloc path. Specifically in `irallocx_prof_sample`: When you reach line 3310, if all of the following are true:
- `old_usize >= SC_LARGE_MINCLASS`
-  `usize <= SC_SMALL_MAXCLASS`
- `bumped_usize >= SC_LARGE_MINCLASS`
Then it's possible the reallocation will be done in-place via `large_ralloc_no_move`. This causes the assertion in `arena_prof_demote` to fail, because the allocation is larger than strictly necessary. I think all we need to do is change the comparison in the assert on line 1264 of arena.c from `==` to `>=`, since this situation _should_ be allowed.

Reply:
Not sure I follow. In the case you described, doesn't it go through `arena_ralloc_move_helper` to realloc+move?

Reply:
No (for context here `LG_PAGE` is 16 so `SC_SMALL_MAXCLASS` is 229376, and cache oblivious is enabled):

<img width="2019" alt="Screenshot 2023-06-13 at 3 22 10 PM" src="https://github.com/jemalloc/jemalloc/assets/35482043/7388955a-6c4f-477e-ad17-f12633834edd">

Reply:
This path indicates:
```
if (oldsize >= SC_LARGE_MINCLASS
	    && usize >= SC_LARGE_MINCLASS) {
```
Is that the case you meant before?

Reply:
Oh I see. The usize here is the bumped_usize, right?

Reply:
I can't repro this locally. It looks CI failures are on FreeBSD?

Does the issue only show up with `cache-oblivious`?

Reply:
NVM I'm able to repro now. So it's indeed only affected by the extra page from cache-oblivious. Let me look more into it.

Reply:
Okay it looks the sz_large_pad is referenced a bit more than I thought, here's what I think we should do first:
1) revert the changes around `pad_size` and `arena_cache_oblivious_randomize`
2) then for the assertion in `arena_prof_demote`, account for the extra page, e.g. bumped_usize = sa2u + sz_large_pad
3) similarly, in the new prof_small test, account for the extra page as well, e.g. expect_zu_eq(edata_size_get(edata), PAGE_CEILING(size) + sz_large_pad)

This would make sure the all "large" alloc still have the extra padding page, even for the sampled / aligned small ones. It's not ideal since it still has 1 extra page that's unnecessary, but we probably should change that separately, given the PR is getting complex already.

Reply:
This also has issues. I feel that it's worth pointing out that all of the pain we're encountering is a consequence of using `size_t bumped_usize = edata_size_get(edata)` on line 1262 of arena.c, instead of the [earlier logic seen here](https://github.com/jemalloc/jemalloc/pull/2459#discussion_r1227285879) (though I understand this approach has better performance characteristics). In this case, the issue is that we're getting failures during stats consolidation `assert(nmalloc >= ndalloc);` on line 141 of arena.c. This appears to be because we're attributing the `alloc` and the `dalloc` for the same sampled profile to different size-classes as a direct consequence of using `size_t bumped_usize = edata_size_get(edata)`.  It's a different manifestation of the same original problem mentioned above, where we shrink the allocation in-place by reusing the existing extent which happens to be large enough. I don't think this works fundamentally.

Reply:
I see. Good point. If we switch back to sa2u to compute bumped usize, does that avoid the troubles? If so let's do that instead. Maybe then assert the edata_size is at most 1 page away with cache_oblivious.

Reply:
In arena_prof_demote. It should hold, right?

Reply:
This does appear to work, at least in the test-cases. I find this surprising, I would've thought we'd run into the same realloc issue where a larger than necessary slab was used to satisfy the allocation, causing the assertion to fail.

Reply:
I might not answered the realloc question properly last week; the shrinking small size case should go down this path:
https://github.com/jemalloc/jemalloc/blob/dev/src/arena.c#L1424

In other words, for small sizes, we only keep the original buffer if the size class is the same. For large realloc, we shrink it by splitting the extent when crossing size classes.

Reply:
I see. There is the edge-case of sampling the reallocation of a small allocation which was itself already sampled, but with a requested size such that the size-class is the same before and after reallocation. As the code currently stands, we don't reuse the existing, promoted extent in this case, but instead use a new one. Do you think it's worth special-casing this by adding some additional logic [here in `arena_ralloc`](https://github.com/jemalloc/jemalloc/blob/d1313313101f9df127bba08bf8fd90a849bf3b87/src/arena.c#L1485), or would you rather just let it use a new extent for simplicity's sake?

Reply:
I'd consider that case to be rather uncommon, and probably not worth optimizing specifically. One other factor for not doing that, is we may still change the round-up-to-PAGE approach in the future, e.g. once we figure out the story for different edata types, it would be possible to use to the first item in each slab for samples (which is still page-aligned) to avoid a lot of the special casing.